### PR TITLE
CustomProgram error

### LINKED ISSUE #3085: Program errors too difficult to troubleshoot
#### Problem

`entrypoint` returns a ProgramError, but because it's a one-size-fits-all enum, programs have split into two terrible ways of dealing with it:
* Use errors that lack precision, such as InvalidArgument
* Add their own error enum and then map all of them to ProgramError::GenericError to cross the chasm

#### Proposed Solution

Add `ProgramError::CustomError(u64)`. Add some facilities to easily map between ProgramError and program-specific errors.

### PR DESCRIPTION
**Pull request by CriesofCarrots**
#### Problem
As described succinctly in #3085, ProgramErrors produced at runtime are difficult to troubleshoot because they don't share any program-specific details. 

#### Summary of Changes
Add CustomError type to ProgramError enum, which takes up to 32 bytes of serialized data
Use simple serialization (and convenience macro) to pass current native program error enum types to CustomError.
...will implement CustomError in other native programs in a separate PR

Fixes #3085 
Toward #1993 
@mvines , if you would expand on your last comment on #1993 , I would be happy to include that clean-up here or in a quick follow-up PR.

### CODE REVIEW

File: programs/system/src/lib.rs [hunk 1]
```diff
@@ -96,11 +98,10 @@ pub fn entrypoint(
                 program_id,
             } => create_system_account(keyed_accounts, lamports, space, &program_id).map_err(|e| {
                 match e {
-                    SystemError::AccountAlreadyInUse => ProgramError::InvalidArgument,
                     SystemError::ResultWithNegativeLamports => {
                         ProgramError::ResultWithNegativeLamports
                     }
-                    SystemError::SourceNotSystemAccount => ProgramError::InvalidArgument,
+                    e => ProgramError::CustomError(custom_error!(e)),
```
Comment (garious) [inline programs/system/src/lib.rs]: I'd prefer this was just:

```rust
e => ProgramError::CustomError(serialize(&e).unwrap()),
```

And that the length check be enforced in runtime.rs. Also, runtime.rs should have a unit-test for that check and resulting truncation.

**Comment by mvines** [comment]
> if you would expand on your last comment on #1993 

This was referring to how most of the fields in `ProgramError` are not actually generated by the program itself but by the runtime, eg search `runtime/src/runtime.rs ` for `ProgramError`.   One possible way out of this;
1. Rename `ProgramError` to `RuntimeError`
2. Rename `ProgramError::CustomError` to  `RuntimeError::ProgramError`.
3. Modify `solana_entrypoint` to only return a `Vec<u8>` that gets wrapped in a `RuntimeError::ProgramError` by the runtime.

**Comment by garious** [comment]
@mvines, one concern there is that some program errors really ought to be the same across all programs (anything to do with serialization or signatures).

I think ProgramError can continue to exist, but several of its errors should be moved to your proposed RuntimeError. But if we go that route, that'd be outside the scope of this PR.

**Comment by mvines** [comment]
Yeah having a small set of standardized ProgramErrors would be tolerable, that's part of the scope of #1993

**Comment by solana-grimes** [comment]
:broken_heart: Unable to automerge due to CI failure

**Pull request merged**